ninjatracing

```diff
@@ -76,17 +81,78 @@ class Threads:
         return len(self.workers) - 1
 
 
-def log_to_dicts(log, pid, show_all):
+def read_events(trace, options):
+    """Reads all events from time-trace json file |trace|."""
+    trace_data = json.load(trace)
+
+    def include_event(event, options):
+        """Only include events if they are complete events, are longer than
+        granularity, and are not totals."""
+        return ((event['ph'] == 'X') and
+                (event['dur'] >= options['granularity']) and
+                (not event['name'].startswith('Total')))
+
+    return [x for x in trace_data['traceEvents'] if include_event(x, options)]
+
+
+def trace_to_dicts(target, trace, options, pid, tid):
+    """Read a file-like object |trace| containing -ftime-trace data and yields
+    about:tracing dict per eligible event in that log."""
+    for event in read_events(trace, options):
+        # Check if any event duration is greater than the duration from ninja.
+        ninja_time = (target.end - target.start) * 1000
+        if event['dur'] > ninja_time:
+            print("Inconsistent timing found (clang time > ninja time). Please"
+                  " ensure that timings are from consistent builds.")
+            sys.exit(1)
+
+        # Set tid and pid from ninja log.
+        event['pid'] = pid
+        event['tid'] = tid
+
+        # Offset trace time stamp by ninja start time.
+        event['ts'] += (target.start * 1000)
+
+        yield event
+
+
+def embed_time_trace(ninja_log_dir, target, pid, tid, options):
+    """Produce time trace output for the specified ninja target. Expects
+    time-trace file to be in .json file named based on .o file."""
+    for t in target.targets:
+        o_path = os.path.join(ninja_log_dir, t)
+        json_trace_path = os.path.splitext(o_path)[0] + '.json'
+        try:
+            with open(json_trace_path, 'r') as trace:
+                for time_trace_event in trace_to_dicts(target, trace, options,
+                                                       pid, tid):
+                    yield time_trace_event
+        except IOError:
+            pass
+
+
+def log_to_dicts(log, pid, options):
     """Reads a file-like object |log| containing a .ninja_log, and yields one
     about:tracing dict per command found in the log."""
     threads = Threads()
-    for target in read_targets(log, show_all):
+    for target in read_targets(log, options['showall']):
+        tid = threads.alloc(target)
+
         yield {
             'name': '%0s' % ', '.join(target.targets), 'cat': 'targets',
             'ph': 'X', 'ts': (target.start * 1000),
             'dur': ((target.end - target.start) * 1000),
-            'pid': pid, 'tid': threads.alloc(target), 'args': {},
+            'pid': pid, 'tid': tid, 'args': {},
             }
+        if options.get('embed_time_trace', False):
```

Comment:
nit: argparse option objects are usually used like `options.embed_time_trace` -- since the flag has a default already, no need for the get() call with the explicit default.

Reply:
I did that to support the test, so I didn't have to add an "embed_time_trace: false" entry to log_to_dicts tests that weren't concerned with this.

Maybe a dictionary of default options in the test would have been better.

Reply:
Ah, makes sense. Doesn't matter all that much then :)